### PR TITLE
[Feature]introduce the data type of Function to support lambda function

### PR DESCRIPTION
**Pull request by fzhedu**
Signed-off-by: fzhedu <fzhedu@gmail.com>

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

add a new data type of Function to support Lambda function , so that the signature of a lambda function can be presented as `Function`. e.g., transform(Array, `Function`).

### CODE REVIEW

**Comment by dulong41** [comment]
title should be marked with their type， like Bugfix, Enhancement, etc.

**Review by satanson — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: .gitignore [hunk 1]
```diff
@@ -69,3 +69,5 @@ be/tags
 !.clang-format
 !.github
 build/
+cmake-build-debug/
+CMakeLists.txt
```
Comment (kangkaisen) [inline .gitignore:73]: Why ignore CMakeLists.txt?

File: be/src/runtime/primitive_type.h [hunk 2]
```diff
@@ -81,6 +82,10 @@ inline bool is_enumeration_type(PrimitiveType type) {
 }
 
 inline bool is_type_compatible(PrimitiveType lhs, PrimitiveType rhs) {
+    if (lhs == TYPE_FUNCTION || rhs == TYPE_FUNCTION) {
```
Comment (kangkaisen) [inline be/src/runtime/primitive_type.h:85]: What if lhs and rhs are both TYPE_FUNCTION?

File: .gitignore (hunk 1 above)
Comment (fzhedu) [reply .gitignore:73]: The file at the root folder just is added to index files for Clion. ignoring it avoids adding it to PRs by accident.

File: be/src/runtime/primitive_type.h (hunk 2 above)
Comment (fzhedu) [reply be/src/runtime/primitive_type.h:85]: Function data type shouldn't be operators for any operations, so making it invalid in any case.

**Comment by fzhedu** [comment]
run starrocks_fe_unittest

**Review by silverbullet233 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 21 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|3|5|60.00%|[257, 465]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|14|85.71%|[650, 1261]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|2|2|100.00%|[]|

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 23 (82.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PrimitiveType.java|3|5|60.00%|[257, 465]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|14|85.71%|[650, 1261]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ScalarType.java|2|2|100.00%|[]|

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by fzhedu** [comment]
merge

**Review by stdpain — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Comment by fzhedu** [comment]
merge

**Pull request merged**